File: lesson_12/structs_ts/src/lesson12.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
import { ListNode } from './list_node.js';

export class Lesson12 {
  push(scores: number): void {
    throw new Error('Not implemented');
  }
  /**
   * Provide the solution to LeetCode 3062 here:
   * https://github.com/yang-su2000/Leetcode-algorithm-practice/tree/master/3062-winner-of-the-linked-list-game
   */
  public gameResult(head: ListNode | null): string {
    if (!head) return 'Tie';

    let current: ListNode | null = head;
    const scores: number[] = [0, 0];

    while (current && current.next) {
      if (current.val !== current.next.val && current.val % 2 == 0) {
        if (current.val > current.next.val) {
          scores[0]++;
        } else {
          scores[1]++;
        }
      }
      current = current.next;
    }
    if (scores[0] > scores[1]) {
      return 'Even';
    } else if (scores[0] < scores[1]) {
      return 'Odd';
    } else {
      return 'Tie';
    }
  }
}

File: lesson_12/structs_java/structs_app/src/main/java/com/codedifferently/lesson12/Lesson12.java
package com.codedifferently.lesson12;


import java.util.ArrayList;
import java.util.List;

public class Lesson12 {

  /**
   * Provide the solution to LeetCode 3062 here:
   * https://github.com/yang-su2000/Leetcode-algorithm-practice/tree/master/3062-winner-of-the-linked-list-game
   */
  

    public String gameResult(ListNode head) {
      List<String> result = new ArrayList<>();
      int evenCounter = 0;
      int oddCounter = 0;
      ListNode current = head;

      while (current != null && current.next != null) {
        int evenValue = current.data;
        int oddValue = current.next.data;

        if (evenValue > oddValue) {
          result.add("even");
          evenCounter++;
        } else {
          result.add("odd");
          oddCounter++;
        }

        current = current.next.next;
      }

      String winningTeam;
      if (evenCounter > oddCounter) {
        winningTeam = "even";
      } else if (oddCounter > evenCounter) {
        winningTeam = "odd";
      } else {
        winningTeam = "tie";
      }

      System.out.println("Node team with higher value in each pair: " + result);
      System.out.println("The team with the most high-values is: " + winningTeam);

      return winningTeam;
    }
  }


